client/src/components/BulkCreateForm/BulkCreateForm.js

```diff
@@ -16,58 +16,154 @@ export default class BulkCreateForm extends React.Component {
+
+        // Reformat the mapping into the Map<String, String> format the backend uses
+        let map = {};
+        for (let row of mapping) {
+            if (!row.mappedField) continue;
+            map[row.originalField] = row.mappedField;
+        }
 
-        this.resetForm(false);
+        const form = {
+            csv: fileUpload.csv,
+            zip: fileUpload.zip,
+            mapping: JSON.stringify(map)
+        };
 
-        if (this.validateForm()) {
-            this.submitForm();
+        if (fileUpload.zip) {
+            // When dealing with zip files, we set the csv to JSZip's object for the CSV file so that we can read the
+            // headers, but it's not the real JavaScript File object, and it isn't useful on the backend, so we don't
+            // want to include it in the request
+            delete form.csv;
         }
+        return form;
+    }
+
+    submitForValidation() {
+        const { onValidationSubmit } = this.props;
+        onValidationSubmit(this.buildForm());
+    }
+
+    submitCreate() {
+        const { onCreateSubmit } = this.props;
+        onCreateSubmit(this.buildForm());
     }
 
     /**
      * Handles when a file is uploaded to the file upload input
      * @param event - The upload event fired by the input
      */
-    handleFileUploadChange(event) {
+    async handleFileUploadChange(event) {
         const { fileUpload } = this.state;
 
         this.resetForm(false);
 
         if (validator.matches(event.target.value, csvFileRegex)) {
-            fileUpload.file = event.target.files[0];
-            fileUpload.message = event.target.files[0].name;
-            fileUpload.fileType = '.csv';
+            fileUpload.csv = event.target.files[0];
             this.setState({fileUpload});
+            this.readCSVHeaders();
         }
         else if (validator.matches(event.target.value, zipFileRegex)) {
-            fileUpload.file = event.target.files[0];
-            fileUpload.message = event.target.files[0].name;
-            fileUpload.fileType = '.zip';
+            fileUpload.zip = event.target.files[0];
+            let content = await JSZip.loadAsync(fileUpload.zip);
+
+            for (let fileName in content.files) {
+                if (!content.files.hasOwnProperty(fileName) || fileName.startsWith('__MACOSX')) continue;
+                if (fileName.endsWith('.csv')) {
+                    if (fileUpload.csv) {
+                        fileUpload.errorMessage = 'Your .zip file contained multiple .csv files. Please only upload one .csv file at a time.';
+                        fileUpload.csv = null;
+                        fileUpload.zip = null;
+                        fileUpload.images = [];
+                        fileUpload.isValid = false;
+                        break;
+                    }
+                    fileUpload.csv = content.files[fileName];
+                } else {
+                    fileUpload.images.push(content.files[fileName]);
```

Comment:
I haven't finished reading the PR, but is assuming that all other files are valid image files going to be an issue somewhere? (Maybe I'll answer my own question as I keep reading).

Reply:
It only really matters if a row in the csv lists that filename as an image. Then the backend will check if it's the right file type.